exporters/logging-otlp/src/test/java/io/opentelemetry/exporter/logging/otlp/AbstractOtlpJsonLoggingExporterTest.java

```diff
@@ -0,0 +1,294 @@
+      Resource.create(Attributes.builder().put("key", "value").build());
+
+  private static final ByteArrayOutputStream STREAM = new ByteArrayOutputStream();
+  private static final PrintStream PRINT_STREAM = new PrintStream(STREAM);
+
+  @RegisterExtension LogCapturer logs;
+  private final String defaultConfigString;
+  private final Class<?> providerClass;
+
+  private final Class<?> componentProviderType;
+
+  private final String expectedFileNoWrapper;
+  private final String expectedFileWrapper;
+
+  public AbstractOtlpJsonLoggingExporterTest(
+      Class<?> exporterClass,
+      Class<?> providerClass,
+      Class<?> componentProviderType,
+      String expectedFileNoWrapper,
+      String expectedFileWrapper,
+      String defaultConfigString) {
+    this.providerClass = providerClass;
+    this.componentProviderType = componentProviderType;
+    this.expectedFileNoWrapper = expectedFileNoWrapper;
+    this.expectedFileWrapper = expectedFileWrapper;
+    logs = LogCapturer.create().captureForType(exporterClass);
+    this.defaultConfigString = defaultConfigString;
+  }
+
+  protected abstract T createExporter(
+      @Nullable OutputStream outputStream, boolean wrapperJsonObject);
+
+  protected abstract T createDefaultExporter();
+
+  protected abstract T createDefaultStdoutExporter();
+
+  protected abstract T toBuilderAndBack(T exporter);
+
+  protected abstract CompletableResultCode export(T exporter);
+
+  protected abstract CompletableResultCode flush(T exporter);
+
+  protected abstract CompletableResultCode shutdown(T exporter);
+
+  private String output(@Nullable OutputStream outputStream) {
+    if (outputStream == null) {
+      return logs.getEvents().stream()
+          .map(LoggingEvent::getMessage)
+          .reduce("", (a, b) -> a + b + "\n")
+          .trim();
+    }
+
+    try {
+      return STREAM.toString(StandardCharsets.UTF_8.name());
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @BeforeAll
+  @SuppressWarnings("SystemOut")
+  static void setUpStatic() {
+    systemOut = System.out;
+    System.setOut(PRINT_STREAM);
+  }
+
+  @AfterAll
+  @SuppressWarnings("SystemOut")
+  static void tearDownStatic() {
+    System.setOut(systemOut);
+  }
+
+  @BeforeEach
+  void setUp() {
+    STREAM.reset();
+  }
+
+  public static class TestCase {
+    @Nullable private final OutputStream outputStream;
+    private final boolean wrapperJsonObject;
+
+    public TestCase(@Nullable OutputStream outputStream, boolean wrapperJsonObject) {
+      this.outputStream = outputStream;
+      this.wrapperJsonObject = wrapperJsonObject;
+    }
+
+    @Nullable
+    public OutputStream getOutputStream() {
+      return outputStream;
+    }
+
+    public boolean isWrapperJsonObject() {
+      return wrapperJsonObject;
+    }
+  }
+
+  @SuppressWarnings("SystemOut")
+  public static Stream<Arguments> exportTestCases() {
+    return Stream.of(System.out, null)
```

Comment:
Let's enumerate all these options instead of multiple layers of `Stream.of().flatmap`. For test code, some duplication is ok to improve readability.